File: src/close_wiki/orchestrator/run_digest.py
"""Main orchestration pipeline for `close-wiki scan`.

Flow:
    1. Snapshot repo → list[FileManifest]
    2. Shard files → list[Shard]
    3. For each shard → run extractor (Docker or local) → AnalysisResult
    4. Persist symbols / relationships to SQLite
    5. Synthesise wiki pages via LLM (host-side)
    6. Build Mermaid diagrams
    7. Export markdown + JSON
"""
from __future__ import annotations

import uuid
from pathlib import Path
from typing import Callable

from close_wiki.models.contracts import AnalysisResult, LLMConfig
from close_wiki.orchestrator.sharding import ShardPlanner
from close_wiki.orchestrator.snapshotter import Snapshotter
from close_wiki.sandbox.runner import BaseRunner, get_runner
from close_wiki.storage.sqlite_store import SqliteStore


def run_digest(
    repo_root: Path,
    output_dir: Path,
    llm_config: LLMConfig | None = None,
    *,
    force_local: bool = False,
    progress: Callable[[str], None] | None = None,
) -> None:
    """Full scan pipeline.

    Args:
        repo_root: Absolute path to the repository to scan.
        output_dir: `.close-wiki/` directory; DB + wiki output goes here.
        llm_config: LLM settings; defaults to LLMConfig() which uses ollama/llama4.
        force_local: Skip Docker even if available.
        progress: Optional callback that receives status strings for display.
    """
    llm_config = llm_config or LLMConfig()
    _log = progress or (lambda _: None)

    run_id = str(uuid.uuid4())
    db_path = output_dir / "store.db"

    store = SqliteStore(db_path)
    store.open()

    try:
        store.upsert_run(run_id, str(repo_root))
        _log(f"Run {run_id[:8]} started")

        # ── 1. Snapshot ──────────────────────────────────────────────
        _log("Snapshotting repository…")
        snapshotter = Snapshotter(repo_root)
        files = snapshotter.snapshot()
        _log(f"  {len(files)} files found")

        store.upsert_snapshot(run_id, [f.model_dump() for f in files])
        for f in files:
            store.upsert_file(run_id, f.path, f.sha256, f.size_bytes, f.language)

        # ── 2. Shard ─────────────────────────────────────────────────
        _log("Planning shards…")
        planner = ShardPlanner()
        shards = planner.plan(files, llm_config)
        _log(f"  {len(shards)} shards")

        # ── 3. Extract ───────────────────────────────────────────────
        runner: BaseRunner = get_runner(force_local=force_local)
        runner_name = type(runner).__name__
        _log(f"Using {runner_name}")

        merged_results: list[AnalysisResult] = []
        for i, shard in enumerate(shards, 1):
            _log(f"  Extracting shard {i}/{len(shards)}: {shard.shard_id}")
            result = runner.run(shard, repo_root)
            merged_results.append(result)

            # Persist
            symbols_dicts = [s.model_dump() for s in result.symbols]
            rels_dicts = [r.model_dump(by_alias=True) for r in result.relationships]
            store.upsert_symbols(run_id, symbols_dicts)
            store.upsert_relationships(run_id, rels_dicts)

        # ── 4. Synthesise wiki pages ─────────────────────────────────
        _log("Synthesising wiki pages…")
        from close_wiki.synthesis.page_builder import PageBuilder  # noqa: PLC0415
        from close_wiki.synthesis.diagram_builder import DiagramBuilder  # noqa: PLC0415

        all_symbols_raw = store.get_all_symbols(run_id)
        print(f"Total symbols: {len(all_symbols_raw)}")
        all_rels_raw = store.get_all_relationships(run_id)

        combined = _combine_results(merged_results)
        builder = PageBuilder(llm_config)
        pages = builder.build(combined)

        diagram_builder = DiagramBuilder()
        diagrams = diagram_builder.build(all_rels_raw)

        for slug, (title, content) in pages.items():
            store.upsert_page(run_id, slug, title, content)
        for name, (dtype, content) in diagrams.items():
            store.upsert_diagram(run_id, name, dtype, content)

        # ── 5. Export ────────────────────────────────────────────────
        _log("Exporting…")
        from close_wiki.exporters.markdown_export import MarkdownExporter  # noqa: PLC0415
        from close_wiki.exporters.json_export import JsonExporter  # noqa: PLC0415

        MarkdownExporter(output_dir).export(pages, diagrams)
        JsonExporter(output_dir).export(run_id, files, combined, pages, diagrams)

        store.update_run_status(run_id, "success")
        _log("Done.")

    except Exception:
        store.update_run_status(run_id, "failed")
        raise
    finally:
        store.close()


def _combine_results(results: list[AnalysisResult]) -> AnalysisResult:
    if not results:
        return AnalysisResult(shard_id="all", files_seen=[], entry_points=[])

    combined = AnalysisResult(shard_id="all", files_seen=[], entry_points=[])
    for r in results:
        combined.files_seen.extend(r.files_seen)
        combined.entry_points.extend(r.entry_points)
        combined.symbols.extend(r.symbols)
        combined.relationships.extend(r.relationships)
        combined.build_commands.extend(r.build_commands)
        combined.test_commands.extend(r.test_commands)
        combined.risks.extend(r.risks)
        combined.evidence.update(r.evidence)
    return combined

File: go/internal/orchestrator/run_digest.go
// Package orchestrator — RunDigest is the full scan pipeline.
package orchestrator

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/fatih/color"
	"github.com/google/uuid"
	"github.com/schollz/progressbar/v3"
	"golang.org/x/sync/errgroup"

	"github.com/unrealandychan/close-wiki/internal/extractor"
	"github.com/unrealandychan/close-wiki/internal/llm"
	"github.com/unrealandychan/close-wiki/internal/models"
	"github.com/unrealandychan/close-wiki/internal/storage"
	"github.com/unrealandychan/close-wiki/internal/synthesis"
)

const maxShardWorkers = 4

// DigestOptions configures a full scan run.
type DigestOptions struct {
	LLMConfig models.LLMConfig
	Verbose   bool
	Progress  func(string) // optional progress callback
	Languages []string     // nil = all languages
}

// RunDigest executes the full scan → plan → generate → export pipeline.
//
// Flow:
//  1. Snapshot repo
//  2. Shard files
//  3. Extract shards in parallel
//  4. Build diagrams
//  5. Plan wiki structure (LLM)
//  6. Generate wiki pages (LLM, concurrent)
//  7. Persist & export
func RunDigest(ctx context.Context, repoRoot, outputDir string, opts DigestOptions) error {
	log := newProgressLogger(opts.Progress, opts.Verbose)

	runID := uuid.New().String()
	dbPath := filepath.Join(outputDir, "store.db")

	store, err := storage.Open(dbPath)
	if err != nil {
		return fmt.Errorf("open store: %w", err)
	}
	defer store.Close()

	if err := store.UpsertRun(runID, repoRoot); err != nil {
		return fmt.Errorf("upsert run: %w", err)
	}
	log.logf("Run %s started", runID[:8])

	// ── 1. Snapshot ────────────────────────────────────────────────────────
	log.logf("Snapshotting repository…")
	snapper := NewSnapshotter(repoRoot, nil, opts.Languages)
	files, err := snapper.Snapshot()
	if err != nil {
		return fmt.Errorf("snapshot: %w", err)
	}
	log.logf("  %d files found", len(files))

	if err := store.UpsertSnapshot(runID, files); err != nil {
		return fmt.Errorf("store snapshot: %w", err)
	}

	// ── 2. Shard ───────────────────────────────────────────────────────────
	log.logf("Planning shards…")
	planner := NewShardPlanner(0)
	shards := planner.Plan(files)
	log.logf("  %d shards planned", len(shards))

	// ── 3. Extract shards in parallel ──────────────────────────────────────
	log.logf("Extracting shards (up to %d parallel)…", maxShardWorkers)
	mergedResults := make([]models.AnalysisResult, len(shards))
	var resultsMu sync.Mutex

	sem := make(chan struct{}, maxShardWorkers)
	eg, egCtx := errgroup.WithContext(ctx)

	reg := extractor.NewRegistry()
	bar := newShardBar(len(shards))

	for i, shard := range shards {
		i, shard := i, shard
		eg.Go(func() error {
			sem <- struct{}{}
			defer func() { <-sem }()

			result, err := extractShard(egCtx, repoRoot, shard, reg)
			resultsMu.Lock()
			defer resultsMu.Unlock()
			_ = bar.Add(1)
			if err != nil {
				log.logf("  ⚠ shard %s failed: %v", shard.ShardID, err)
				mergedResults[i] = models.AnalysisResult{
					ShardID:     shard.ShardID,
					FilesSeen:   []string{},
					EntryPoints: []string{},
					Risks:       []string{fmt.Sprintf("extraction failed: %v", err)},
				}
				return nil // graceful — don't abort the whole run
			}
			mergedResults[i] = result
			log.logf("  ✓ %s: %d symbols, %d rels", shard.ShardID, len(result.Symbols), len(result.Relationships))
			return nil
		})
	}
	if err := eg.Wait(); err != nil {
		return err
	}

	// Persist results (SQLite writes must be serial)
	for _, result := range mergedResults {
		if err := store.UpsertSymbols(runID, result.Symbols); err != nil {
			return fmt.Errorf("store symbols: %w", err)
		}
		if err := store.UpsertRelationships(runID, result.Relationships); err != nil {
			return fmt.Errorf("store rels: %w", err)
		}
	}

	// ── 4. Combine + build diagrams ────────────────────────────────────────
	log.logf("Building diagrams…")
	combined := combineResults(mergedResults)
	db := synthesis.NewDiagramBuilder()
	diagrams := db.Build(combined.Relationships, combined.EntryPoints)
	log.logf("  %d diagram(s) built", len(diagrams))

	if mg, ok := diagrams["module-graph"]; ok {
		combined.Evidence["pre_built_module_graph"] = mg[1]
	}
	if ch, ok := diagrams["class-hierarchy"]; ok {
		combined.Evidence["pre_built_dependency_graph"] = ch[1]
	}

	// ── 5. Plan wiki structure ─────────────────────────────────────────────
	log.logf("Planning wiki structure…")
	llmClient := llm.New(opts.LLMConfig)
	wikiPlanner := synthesis.NewPlannerAgent(llmClient)
	plan, err := wikiPlanner.Plan(ctx, combined)
	if err != nil {
		return fmt.Errorf("plan wiki: %w", err)
	}
	log.logf("  %d pages planned", len(plan.Pages))

	// ── 6. Generate wiki pages ─────────────────────────────────────────────
	log.logf("Generating wiki pages…")
	pageBuilder := synthesis.NewPageBuilder(llmClient)
	diagMap := make(map[string][2]string)
	for k, v := range diagrams {
		diagMap[k] = [2]string(v)
	}
	pages, err := pageBuilder.BuildAll(ctx, plan, combined, diagMap)
	if err != nil {
		return fmt.Errorf("build pages: %w", err)
	}
	log.logf("  %d pages generated", len(pages))

	// ── 7. Persist pages & export ──────────────────────────────────────────
	log.logf("Persisting pages…")
	for slug, content := range pages {
		title := slug
		for _, spec := range plan.Pages {
			if spec.Slug == slug {
				title = spec.Title
				break
			}
		}
		if err := store.UpsertPage(runID, slug, title, content); err != nil {
			return fmt.Errorf("store page %q: %w", slug, err)
		}
	}

	// Export to disk
	wikiDir := filepath.Join(outputDir, "wiki")
	if err := os.MkdirAll(wikiDir, 0o755); err != nil {
		return fmt.Errorf("mkdir wiki: %w", err)
	}
	for slug, content := range pages {
		path := filepath.Join(wikiDir, slug+".md")
		if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
			return fmt.Errorf("write page %q: %w", slug, err)
		}
	}

	log.logf("✅ Done — run %s | %d pages in %s", runID[:8], len(pages), outputDir)
	return nil
}

// ── helpers ───────────────────────────────────────────────────────────────────

func extractShard(_ context.Context, repoRoot string, shard models.Shard, reg *extractor.Registry) (models.AnalysisResult, error) {
	if reg == nil {
		reg = extractor.NewRegistry()
	}
	result := models.AnalysisResult{
		ShardID:   shard.ShardID,
		FilesSeen: make([]string, 0, len(shard.Files)),
		Evidence:  make(map[string]string),
	}

	for _, fm := range shard.Files {
		fullPath := filepath.Join(repoRoot, fm.Path)
		ext := filepath.Ext(fm.Path)
		ar := reg.ExtractFile(fullPath, fm.Path, ext)
		result.FilesSeen = append(result.FilesSeen, fm.Path)
		result = mergeInto(result, ar)
	}
	return result, nil
}

func mergeInto(base, extra models.AnalysisResult) models.AnalysisResult {
	base.Symbols = append(base.Symbols, extra.Symbols...)
	base.Relationships = append(base.Relationships, extra.Relationships...)
	base.EntryPoints = append(base.EntryPoints, extra.EntryPoints...)
	base.BuildCommands = append(base.BuildCommands, extra.BuildCommands...)
	base.TestCommands = append(base.TestCommands, extra.TestCommands...)
	base.Risks = append(base.Risks, extra.Risks...)
	for k, v := range extra.Evidence {
		base.Evidence[k] = v
	}
	return base
}

func combineResults(results []models.AnalysisResult) models.AnalysisResult {
	combined := models.AnalysisResult{Evidence: make(map[string]string)}
	for _, r := range results {
		combined.FilesSeen = append(combined.FilesSeen, r.FilesSeen...)
		combined.Symbols = append(combined.Symbols, r.Symbols...)
		combined.Relationships = append(combined.Relationships, r.Relationships...)
		combined.EntryPoints = append(combined.EntryPoints, r.EntryPoints...)
		combined.BuildCommands = append(combined.BuildCommands, r.BuildCommands...)
		combined.TestCommands = append(combined.TestCommands, r.TestCommands...)
		combined.Risks = append(combined.Risks, r.Risks...)
		for k, v := range r.Evidence {
			combined.Evidence[k] = v
		}
	}
	return combined
}

// progressLogger emits coloured messages to stderr and an optional callback.
type progressLogger struct {
	cb      func(string)
	verbose bool

	cyan    *color.Color
	green   *color.Color
	yellow  *color.Color
	red     *color.Color
}

func newProgressLogger(cb func(string), verbose bool) *progressLogger {
	return &progressLogger{
		cb:      cb,
		verbose: verbose,
		cyan:    color.New(color.FgCyan, color.Bold),
		green:   color.New(color.FgGreen),
		yellow:  color.New(color.FgYellow),
		red:     color.New(color.FgRed),
	}
}

func (l *progressLogger) logf(format string, args ...any) {
	msg := fmt.Sprintf(format, args...)
	if l.cb != nil {
		l.cb(msg)
	}
	// Colour by content
	switch {
	case strings.HasPrefix(msg, "⚠") || strings.HasPrefix(msg, "  ⚠"):
		l.yellow.Fprintln(os.Stderr, msg)
	case strings.HasPrefix(msg, "  ✓") || strings.HasPrefix(msg, "✅"):
		l.green.Fprintln(os.Stderr, msg)
	case strings.HasPrefix(msg, "Error") || strings.HasPrefix(msg, "error"):
		l.red.Fprintln(os.Stderr, msg)
	default:
		l.cyan.Fprintln(os.Stderr, msg)
	}
}

// newShardBar creates a progressbar for shard processing.
func newShardBar(total int) *progressbar.ProgressBar {
	return progressbar.NewOptions(total,
		progressbar.OptionSetDescription("  extracting shards"),
		progressbar.OptionSetWriter(os.Stderr),
		progressbar.OptionShowCount(),
		progressbar.OptionSetTheme(progressbar.Theme{
			Saucer:        "=",
			SaucerHead:    ">",
			SaucerPadding: " ",
			BarStart:      "[",
			BarEnd:        "]",
		}),
		progressbar.OptionClearOnFinish(),
	)
}
